indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/LagBasedAutoScaler.java

```diff
@@ -0,0 +1,242 @@
+    this.lagMetricsQueue = new CircularFifoQueue<>(slots);
+    this.allocationExec = Execs.scheduledSingleThreaded(StringUtils.encodeForFormat(supervisorId) + "-Allocation-%d");
+    this.lagComputationExec = Execs.scheduledSingleThreaded(StringUtils.encodeForFormat(supervisorId) + "-Computation-%d");
+    this.spec = spec;
+    this.supervisor = supervisor;
+  }
+
+  @Override
+  public void start()
+  {
+    Callable<Integer> scaleAction = () -> {
+      LOCK.lock();
+      int desiredTaskCount = -1;
+      try {
+        desiredTaskCount = computeDesiredTaskCount(new ArrayList<>(lagMetricsQueue));
+
+        if (desiredTaskCount != -1) {
+          lagMetricsQueue.clear();
+        }
+      }
+      catch (Exception ex) {
+        log.warn(ex, "Exception while computing desired task count for [%s]", dataSource);
+      }
+      finally {
+        LOCK.unlock();
+      }
+      return desiredTaskCount;
+    };
+
+    lagComputationExec.scheduleAtFixedRate(
+        computeAndCollectLag(),
+        lagBasedAutoScalerConfig.getScaleActionStartDelayMillis(), // wait for tasks to start up
+        lagBasedAutoScalerConfig.getLagCollectionIntervalMillis(),
+        TimeUnit.MILLISECONDS
+    );
+    allocationExec.scheduleAtFixedRate(
+        supervisor.buildDynamicAllocationTask(scaleAction),
+        lagBasedAutoScalerConfig.getScaleActionStartDelayMillis() + lagBasedAutoScalerConfig
+            .getLagCollectionRangeMillis(),
+        lagBasedAutoScalerConfig.getScaleActionPeriodMillis(),
+        TimeUnit.MILLISECONDS
+    );
+    log.info(
+        "LagBasedAutoScaler will collect lag every [%d] millis and will keep [%d] data points for the last [%d] millis for dataSource [%s]",
+        lagBasedAutoScalerConfig.getLagCollectionIntervalMillis(), lagMetricsQueue.size(),
+        lagBasedAutoScalerConfig.getLagCollectionRangeMillis(), dataSource
+    );
+  }
+
+  @Override
+  public void stop()
+  {
+    allocationExec.shutdownNow();
+    lagComputationExec.shutdownNow();
+  }
+
+  @Override
+  public void reset()
+  {
+    // clear queue for kafka lags
+    if (lagMetricsQueue != null) {
+      try {
+        LOCK.lock();
+        lagMetricsQueue.clear();
+      }
+      catch (Exception e) {
+        log.warn(e, "Error,when clear queue in rest action");
+      }
+      finally {
+        LOCK.unlock();
+      }
+    }
+  }
+
+  /**
+   * This method computes current consumer lag. Gets the total lag of all partitions and fill in the lagMetricsQueue
+   *
+   * @return a Runnbale object to compute and collect lag.
+   */
+  private Runnable computeAndCollectLag()
+  {
+    return () -> {
+      LOCK.lock();
+      try {
+        if (!spec.isSuspended()) {
+          LagStats lagStats = supervisor.computeLagStats();
+          if (lagStats == null) {
+            lagMetricsQueue.offer(0L);
+          } else {
+            long totalLags = lagStats.getTotalLag();
+            lagMetricsQueue.offer(totalLags > 0 ? totalLags : 0L);
```

Comment:
Because we can occasionally get negative lags in our practice. Something like https://stackoverflow.com/questions/60847952/how-to-get-rid-of-negative-consumer-lag-in-kafka

Negative lag values is un-necessary and a poison into our lag metrics. So just filter it here.